File: Assignment7/collcount.fs
\ Created by Luke Underwood
\ 2021-04-18
\ collcount.fs
\ Assignment 7
\ CS 331
\ contains definition of word collcount, which has stack effect (n -- c), 
\   where n is a positive integer, and c is the number of iterations of the 
\   Collatz function required to take n to 1

\ perform collatz function on n, increment i, leave i on stack in base case
: collatz { n i -- c }
    \ BASE CASE: n == 1, return i
    n 1 = if
        i

    \ RECURSIVE CASE: n != 1, perform collatz and recurse
    else
        i 1 + to i
        n 2 MOD 0= if
            n 2 / to n
            n i recurse
        else
            n 3 * 1 + to n
            n i recurse
        then
    then
;

: collcount { n -- c }
    n 1 < if
        ." Bad input - please try again!"
    else
        n 0 collatz
    then
;

File: Assignment7/collcount.pl
% Created by Luke Underwood
% 2021-04-19
% collcount.pl
% Assignment 7
% CS331
% contains predicate collcount, which correlates a number with the number of
% iterations of the collatz function required to reduce the number to 1

% counter/3
% counter(+n, +s, ?i) 
% i is the number of iterations that have already been performed
% n is the number initially given to collcount
% s is a starting point for i, should always be set to 0 by collcount
counter(1, S, I) :- I is S.
counter(N, S, I) :- 
                N > 1,
                N mod 2 =:= 0,
                S1 is S + 1,
                N1 is N / 2,
                counter(N1, S1, I).
counter(N, S, I) :- 
                N > 1,
                mod(N, 2) > 0,
                S1 is S + 1,
                N1 is N * 3 + 1,
                counter(N1, S1, I).


% collcount/2
% collcount(+n, ?c) c is iterations of collatz for number n
collcount(N, C) :- 
                N >= 1,
                counter(N, 0, C).